Description: Fix stack out-of-bounds NUL write in EF_SMARTZIP handler [CAN-2026-2034443]
Origin: https://salsa.debian.org/sanvila/unzip/-/raw/debian/6.0-30/debian/patches/can-2026-2034443.patch
Bug: https://bugs.debian.org/1142905
Tracker: https://security-tracker.debian.org/tracker/TEMP-1142905

--- a/zipinfo.c
+++ b/zipinfo.c
@@ -1702,9 +1702,12 @@
                     if ((eb_datalen == EB_SMARTZIP_HLEN) &&
                         makelong(ef_ptr) == 0x70695A64 /* "dZip" */) {
                         char filenameBuf[32];
+                        int filename_length = ef_ptr[32] ;
                         zi_showMacTypeCreator(__G__ &ef_ptr[4]);
                         memcpy(filenameBuf, &ef_ptr[33], 31);
-                        filenameBuf[ef_ptr[32]] = '\0';
+                        if (filename_length > 31)
+                            filename_length = 31;
+                        filenameBuf[filename_length] = '\0';
                         A_TO_N(filenameBuf);
                         Info(slide, 0, ((char *)slide,
                              LoadFarString(ZipItFname), filenameBuf));
